File: mil_common/mil_missions/test/test_run_mission.py
#!/usr/bin/env python3
import unittest

import rospy
from actionlib import TerminalState
from mil_missions_core import MissionClient


class RunMissionTest(unittest.TestCase):
    def __init__(self, *args):
        self.client = MissionClient()
        self.client.wait_for_server()
        super().__init__(*args)

    def test_run_mission(self):
        self.client.run_mission("PrintAndWait")
        self.client.wait_for_result()
        result = self.client.get_result()
        state = self.client.get_state()
        self.assertEqual(state, TerminalState.SUCCEEDED)
        self.assertTrue(result.success)
        self.assertEqual(result.parameters, "")
        self.assertEqual(result.result, "The darkness isn't so scary")

    def test_failing_mission(self):
        self.client.run_mission("FailingMission")
        self.client.wait_for_result()
        result = self.client.get_result()
        state = self.client.get_state()
        self.assertEqual(state, TerminalState.ABORTED)
        self.assertFalse(result.success)
        self.assertTrue(result.parameters)
        self.assertRegex(
            result.result, r"^In submission \<.*\>: This is an example error\!$"
        )


if __name__ == "__main__":
    import rostest

    rospy.init_node("run_mission_test")
    rostest.rosrun("mil_missions", "run_mission", RunMissionTest)

File: mil_common/mil_missions/test/test_run_mission.cpp
#include <gtest/gtest.h>
#include <mil_missions/client.h>
#include <ros/ros.h>
#include <iostream>

#include <mil_missions/DoMissionResult.h>
#include <actionlib/client/simple_action_client.h>

using namespace mil_missions;

class MissionClientTest : public ::testing::Test
{
protected:
  MissionClientTest() : client(nh) {}

  ros::NodeHandle nh;
  mil_missions::MissionClient client;
};

TEST_F(MissionClientTest, RunMission)
{
  EXPECT_TRUE(client.wait_for_server());

  // Test running a mission
  client.run_mission("PrintAndWait");
  EXPECT_TRUE(client.wait_for_result(ros::Duration(5.0)));
  mil_missions::DoMissionResult::ConstPtr result = client.get_result();
  actionlib::SimpleClientGoalState state = client.get_state();
  EXPECT_EQ(state, actionlib::SimpleClientGoalState::SUCCEEDED);
  EXPECT_TRUE(result->success);
  EXPECT_EQ(result->parameters, "");
  EXPECT_EQ(result->result, "The darkness isn't so scary");
}

TEST_F(MissionClientTest, FailingMission) {
  EXPECT_TRUE(client.wait_for_server());

  // Test running a mission
  client.run_mission("FailingMission");
  EXPECT_TRUE(client.wait_for_result(ros::Duration(5.0)));
  mil_missions::DoMissionResult::ConstPtr result = client.get_result();
  actionlib::SimpleClientGoalState state = client.get_state();
  EXPECT_EQ(state, actionlib::SimpleClientGoalState::ABORTED);
  EXPECT_FALSE(result->success);
  EXPECT_TRUE(!result->parameters.empty());
  EXPECT_EQ(result->result, "Exception occurred in FailingMission mission: ValueError: This is an example error!");
}

TEST_F(MissionClientTest, CancelledTest) {
    EXPECT_TRUE(client.wait_for_server());
    
    // Test running a mission
    client.run_mission("CancelledMission");
    EXPECT_TRUE(client.wait_for_result(ros::Duration(5.0)));
    mil_missions::DoMissionResult::ConstPtr result = client.get_result();
    actionlib::SimpleClientGoalState state = client.get_state();
    EXPECT_EQ(state, actionlib::SimpleClientGoalState::ABORTED);
    EXPECT_FALSE(result->success);
    EXPECT_TRUE(result->parameters.empty());
    EXPECT_EQ(result->result, "mission cancelled");
}

int main(int argc, char **argv)
{
  testing::InitGoogleTest(&argc, argv);
  ros::init(argc, argv, "run_mission_test_cpp");
  return RUN_ALL_TESTS();
}
